src/internal/components/file-input/index.tsx

```
@@ -0,0 +1,128 @@
+import { joinStrings } from '../../utils/strings';
+import { FileInputProps } from './interfaces';
+
+import styles from './styles.css.js';
+
+export { FileInputProps };
+
+const InternalFileInput = React.forwardRef(
+  (
+    {
+      accept,
+      ariaRequired,
+      ariaLabel,
+      multiple = false,
+      value,
+      onChange,
+      variant = 'button',
+      children,
+      ...restProps
+    }: FileInputProps,
+    ref: Ref<FileInputProps.Ref>
+  ) => {
+    const uploadInputRef = useRef<HTMLInputElement>(null);
+    const uploadButtonLabelId = useUniqueId('upload-button-label');
+    const formFieldContext = useFormFieldContext(restProps);
+    const selfControlId = useUniqueId('upload-input');
+    const controlId = formFieldContext.controlId ?? selfControlId;
+    const baseProps = getBaseProps(restProps);
+
+    useForwardFocus(ref, uploadInputRef);
+
+    const [isFocused, setIsFocused] = useState(false);
+    const onUploadButtonClick = () => uploadInputRef.current?.click();
+    const onUploadInputFocus = () => setIsFocused(true);
+    const onUploadInputBlur = () => setIsFocused(false);
+
+    const onUploadInputChange = ({ target }: ChangeEvent<HTMLInputElement>) => {
+      fireNonCancelableEvent(onChange, { value: target.files ? Array.from(target.files) : [] });
+    };
+
+    checkControlled('FileInput', 'value', value, 'onChange', onChange);
+
+    const nativeAttributes: React.HTMLAttributes<HTMLInputElement> = {
+      'aria-label': ariaLabel || children,
+      'aria-labelledby': joinStrings(formFieldContext.ariaLabelledby, uploadButtonLabelId),
+      'aria-describedby': formFieldContext.ariaDescribedby,
+    };
+    if (formFieldContext.invalid) {
+      nativeAttributes['aria-invalid'] = true;
+    }
+    if (ariaRequired) {
+      nativeAttributes['aria-required'] = true;
+    }
+
+    if (variant === 'icon' && !ariaLabel) {
+      warnOnce('FileInput', 'Aria label is required with icon variant.');
```

For reviewer: is this how we typically enforce a property that is only required in certain configurations? Should we throw an error instead?

---

Discussed with Andrei - this is typical.